[Resolve #623] Catch missing config attributes

devenney opened the pull request:
Signed-off-by: Brendan Devenney <brendan.devenney@cloudreach.com>

There are some unanswered questions to consider when reviewing this PR:

* Should the code be differentiating between Stack Config and Stack Group Config for the user's benefit? I'm a little thrown off by the fact you can successfully deploy a Stack without any files in the tree named `config.yaml`, despite the implication that `config.yaml` is mandatory in the [StackGroup Config](https://sceptre.cloudreach.com/latest/docs/stack_group_config.html) docs. Effectively, you can store all of your config in `vpc.yaml` and Sceptre won't complain.
* Do we care enough about invalid keys to add these checks? My opinion is yes, as a typo in a stack's config could be masked due to the attribute cascading down from a parent directory.

Also, for what it's worth, I have no idea if this is the most efficient way to flatten lists of lists into a single flat list. Some rough testing suggests it's the fasted way I know but please do consider if there's a better/cleaner way. 👍 

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes flake8 (`make lint`) checks.
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

nabeelamjad commented on sceptre/config/reader.py:
```diff
@@ -424,6 +439,29 @@ def _construct_stack(self, rel_path, stack_group_config=None):
+        for required_key in STACK_CONFIG_ATTRIBUTES.required:
+            if required_key not in config:
+                raise InvalidConfigFileError(
+                    "Required key '{0}' not found in Stack Config ({1})".format(
```
Probably want to keep the error message format consistent with the one above.

If we don't care about the _exact_ error message (Stack Config / Stack Group Config) then we should probably use a single loop to check both required config attributes.

I believe you mentioned that the **stack group config** can be missing with everything in the **stack config** instead, which means that the error message above (Stack Config / Stack Group Config) won't be correct anyhow. We could fix the underlying issue to force the correct attributes to exist in the `config.yaml` but that would definitely be a breaking change at this point, so I'd rather suggest we change the error message instead.

devenney commented:
Thanks for the review @nabeelamjad. tl;dr of my latest push:

* I have removed the invalid key testing. Clearly this issue isn't as clear-cut as I had assumed and I feel it warrants its own issue rather than flying under the radar with this fix. I have left the `INTERNAL_CONFIG_ATTRIBUTES` code though as this was key to me understanding the codebase when determining valid keys.
* The check no longer differentiates between Stack Config and StackGroup Config as discussed.
* I've added a test for each mandatory key. I can take a more holistic approach with this to ensure it automatically picks up added/removed required keys if you want but it feels overkill.
* I removed the vim swap file and added vim regexes to `.gitignore`.
